syntax = "proto3";
package putty_interface;

service RemoteConnection {
  rpc CreateRemoteConnection (CreateRequest) returns (ConnectionId);
  rpc Write            (WriteRequest)  returns (Empty);
  rpc Stop             (ConnectionId)  returns (Empty);
  rpc Read             (ConnectionId)  returns (stream ByteChunk);
}

message CreateRequest {
  oneof kind {
    Serial serial = 1;
    Ssh    ssh    = 2;
  }
}

message Serial { string port = 1; uint32 baud = 2; }
message Ssh    { string host = 1; uint32 port = 2; string user = 3; string pass = 4; }

message ConnectionId { string id = 1; }
message WriteRequest { string id = 1; bytes data = 2; }
message ByteChunk    { bytes data = 1; }
message Empty        {}
